### PR TITLE
Search and Replace: Support advanced path expressions

### PR DESCRIPTION
**Pull request by phanimarupaka**
@mortent 

This PR supports more path expressions for search and replace.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/cpuguy83/go-md2man/v2 v2.0.0
 	github.com/go-errors/errors v1.0.1
 	github.com/go-openapi/spec v0.19.5
+	github.com/golangplus/testing v0.0.0-20180327235837-af21d9c3145e
```
Comment (mortent) [inline go.mod:9]: Why are we adding this?

File: internal/util/search/pathparser.go [hunk 2]
```diff
@@ -10,17 +10,61 @@ func (sr *SearchReplace) pathMatch(yamlPath string) bool {
...
+	patternElems := strings.Split(sr.ByPath, PathDelimiter)
+	yamlPathElems := strings.Split(strings.TrimPrefix(yamlPath, PathDelimiter), PathDelimiter)
+	return backTrackMatch(yamlPathElems, patternElems)
+}
+
+// backTrackMatch matches the traversed yamlPathElems with input(from by-path) patternElems
+// * matches any element, ** matches 0 or more elements, array elements are split and matched
+// refer to pathparser_test.go
+func backTrackMatch(yamlPathElems, patternElems []string) bool {
+	yamlPathElemsLen, patternElemsLen := len(yamlPathElems), len(patternElems)
```
Comment (mortent) [inline internal/util/search/pathparser.go:23]: Can we explain with some comments the overall idea about what we are doing in the function? It is pretty hard to follow.

File: internal/cmdsearch/searchreplace_test.go [hunk 3]
```diff
@@ -293,8 +293,38 @@ metadata:
  `,
 	},
 	{
-		name: "search by array objects path",
-		args: []string{"--by-path", "spec.foo[1].c"},
+		name: "search replace by array path regex",
+		args: []string{"--by-path", "spec.foo[1]", "--put-literal", "c"},
```
Comment (mortent) [inline internal/cmdsearch/searchreplace_test.go:297]: What happens if the provided number is outside the bounds of the array?

File: go.mod (hunk 1 above)
Comment (phanimarupaka) [reply go.mod]: By mistake. Removed it.

File: internal/util/search/pathparser.go (hunk 2 above)
Comment (phanimarupaka) [reply internal/util/search/pathparser.go:26]: Done. This is similar to wildcard match https://leetcode.com/problems/wildcard-matching/

File: internal/cmdsearch/searchreplace_test.go (hunk 3 above)
Comment (phanimarupaka) [reply internal/cmdsearch/searchreplace_test.go:297]: It's a no match. Similar to providing non-existing path. We will support appending elements to arrays in future. https://mikefarah.gitbook.io/yq/usage/path-expressions#appending-to-arrays

Comment (mortent) [reply internal/cmdsearch/searchreplace_test.go:297]: This seems like something that would be worthy of testing. This also includes the use of negative indexes (as described in the link) if that is something that we support.

Comment (phanimarupaka) [reply internal/cmdsearch/searchreplace_test.go:297]: Added a [test case](https://github.com/GoogleContainerTools/kpt/pull/1225/files#diff-a0a456fc7bee0fb4aadca84d28d1dd6444127841d2b9499262d1e09cae3b2945R326). We may add more features based on the requirements in future. yq supports negative indices but for us it depends on use-cases we might encounter in the future.

**Review by mortent — APPROVED** [review]
